traversal/selector/exploreRange.go

```diff
@@ -0,0 +1,86 @@
+package selector
+
+import (
+	"fmt"
+
+	ipld "github.com/ipld/go-ipld-prime"
+)
+
+// ExploreRange traverses a list, and for each element in the range specified,
+// will apply a next selector to those reached nodes.
+type ExploreRange struct {
+	next        Selector         // selector for element we're interested in
+	hasSelector map[int]struct{} // a quick check whether to return the selector
+	interest    []PathSegment    // index of element we're interested in
+}
+
+// Interests for ExploreRange are all path segments within the iteration range
+func (s ExploreRange) Interests() []PathSegment {
+	return s.interest
+}
+
+// Explore returns the node's selector if
+// the path matches an index in the range of this selector
+func (s ExploreRange) Explore(n ipld.Node, p PathSegment) Selector {
+	if n.ReprKind() != ipld.ReprKind_List {
+		return nil
+	}
+	index, err := p.Index()
+	if err != nil {
+		return nil
+	}
+	if _, ok := s.hasSelector[index]; !ok {
+		return nil
+	}
+	return s.next
+}
+
+// Decide always returns false because this is not a matcher
+func (s ExploreRange) Decide(n ipld.Node) bool {
+	return false
+}
+
+// ParseExploreRange assembles a Selector
+// from a ExploreRange selector node
+func ParseExploreRange(n ipld.Node) (Selector, error) {
+	if n.ReprKind() != ipld.ReprKind_Map {
+		return nil, fmt.Errorf("selector spec parse rejected: selector body must be a map")
+	}
+	startNode, err := n.TraverseField(startKey)
+	if err != nil {
+		return nil, fmt.Errorf("selector spec parse rejected: start field must be present in ExploreRange selector")
+	}
+	startValue, err := startNode.AsInt()
+	if err != nil {
+		return nil, fmt.Errorf("selector spec parse rejected: start field must be a number in ExploreRange selector")
+	}
+	endNode, err := n.TraverseField(endKey)
+	if err != nil {
+		return nil, fmt.Errorf("selector spec parse rejected: end field must be present in ExploreRange selector")
+	}
+	endValue, err := endNode.AsInt()
+	if err != nil {
+		return nil, fmt.Errorf("selector spec parse rejected: end field must be a number in ExploreRange selector")
+	}
+	if startValue >= endValue {
+		return nil, fmt.Errorf("selector spec parse rejected: end field must be greater than start field in ExploreRange selector")
+	}
+	next, err := n.TraverseField(nextSelectorKey)
+	if err != nil {
+		return nil, fmt.Errorf("selector spec parse rejected: next field must be present in ExploreRange selector")
+	}
+	selector, err := ParseSelector(next)
+	if err != nil {
+		return nil, err
+	}
+	x := ExploreRange{
+		selector,
+		make(map[int]struct{}, endValue-startValue),
+		make([]PathSegment, 0, endValue-startValue),
+	}
+	for i := startValue; i < endValue; i++ {
+		x.interest = append(x.interest, PathSegmentInt{I: i})
+		x.hasSelector[i] = struct{}{}
```

Comment:
Ok, so we've kinda screwed for the interest set on this one by the interface (owch), but I'd definitely go for just storing the two ints and doing the range check again in the Explore method rather than a map.

Reply:
ok. fair. is that a problem for ExploreUnion, which pretty explicitly says "assumes interests isn't a big computation?"

Reply:
This comment -- https://github.com/ipld/go-ipld-prime/pull/17/files/c590adc105460366fffaf07318b86b03f528cb67#diff-0f5c839144c4c664fcef0ebf75f3eb79R25

Reply:
yeah, that's what I mean by "we're kinda screwed for the interest set" -- I agree there's not an obvious way to avoid pregenerating that slice.

The map is pretty easy to ditch though, I think?